File: Big Data/Using Python/Install sklearn, demonstrate classification using KNN, Evaluate confusion matrix and accuracy.py
# Required libraries
import pandas as pd
import numpy as np
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix, ConfusionMatrixDisplay
import matplotlib.pyplot as plt

# ===== Step 1: Load the Dataset =====
# Using Iris dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/iris.csv"
df = pd.read_csv(url)

# --- If file is downloaded locally, use this instead: ---
# df = pd.read_csv("iris.csv")
# --------------------------------------------------------

print("Dataset shape:", df.shape)
print("\nFirst 5 rows:")
print(df.head())
print("\nClass distribution:")
print(df['species'].value_counts())

# ===== Step 2: Prepare Features (X) and Target (y) =====
X = df[['sepal_length', 'sepal_width', 'petal_length', 'petal_width']]
y = df['species']

# ===== Step 3: Split into Training (70%) and Testing (30%) =====
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)
print(f"\nTraining samples: {len(X_train)}")
print(f"Testing samples: {len(X_test)}")

# ===== Step 4: Feature Scaling (Important for KNN - distance based) =====
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# ===== Step 5: Build KNN Classifier =====
k = 5  # Number of neighbors
knn_model = KNeighborsClassifier(n_neighbors=k, metric='euclidean')
knn_model.fit(X_train_scaled, y_train)
print(f"\nKNN Model trained successfully! (K={k})")

# ===== Step 6: Make Predictions =====
y_pred = knn_model.predict(X_test_scaled)

# ===== Step 7: Evaluate Model Accuracy =====
accuracy = accuracy_score(y_test, y_pred)
print(f"\n{'='*50}")
print(f"  MODEL ACCURACY: {accuracy * 100:.2f}%")
print(f"{'='*50}")

print("\nClassification Report:")
print(classification_report(y_test, y_pred))

# ===== Step 8: Confusion Matrix =====
cm = confusion_matrix(y_test, y_pred)
print("Confusion Matrix:")
print(cm)

# Plot Confusion Matrix
disp = ConfusionMatrixDisplay(confusion_matrix=cm, display_labels=knn_model.classes_)
disp.plot(cmap='Blues', values_format='d')
plt.title(f"KNN (K={k}) - Confusion Matrix (Iris Dataset)")
plt.tight_layout()
plt.savefig("knn_confusion_matrix.png")
plt.show()

# ===== Step 9: Find Optimal K =====
print("\nFinding optimal K value...")
k_range = range(1, 21)
accuracies = []

for k in k_range:
    knn = KNeighborsClassifier(n_neighbors=k)
    knn.fit(X_train_scaled, y_train)
    pred = knn.predict(X_test_scaled)
    accuracies.append(accuracy_score(y_test, pred))

# Display accuracy for each K
print(f"\n{'K':<5} {'Accuracy':>10}")
print(f"{'-'*5} {'-'*10}")
for k, acc in zip(k_range, accuracies):
    print(f"{k:<5} {acc*100:>9.2f}%")

best_k = list(k_range)[np.argmax(accuracies)]
print(f"\nBest K = {best_k} with Accuracy = {max(accuracies)*100:.2f}%")

# Plot K vs Accuracy
plt.figure(figsize=(8, 5))
plt.plot(k_range, accuracies, 'bo-', linewidth=2, markersize=8)
plt.axvline(x=best_k, color='red', linestyle='--', label=f'Best K={best_k}')
plt.xlabel("K (Number of Neighbors)")
plt.ylabel("Accuracy")
plt.title("KNN - Accuracy vs K Value")
plt.legend()
plt.grid(True)
plt.tight_layout()
plt.savefig("knn_k_vs_accuracy.png")
plt.show()

# ===== Step 10: Predict for New Data =====
new_flower = pd.DataFrame({'sepal_length': [5.1], 'sepal_width': [3.5],
                           'petal_length': [1.4], 'petal_width': [0.2]})
new_scaled = scaler.transform(new_flower)
prediction = knn_model.predict(new_scaled)
print(f"\nPrediction for new flower (5.1, 3.5, 1.4, 0.2): {prediction[0]}")

File: Big Data/Using R/Install sklearn, demonstrate classification using KNN, Evaluate confusion matrix and accuracy.R
# Required libraries
library(class)    # KNN
library(caret)    # Confusion matrix & accuracy

# ===== Step 1: Load the Dataset =====
url <- "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/iris.csv"
df <- read.csv(url)

# --- If file is downloaded locally, use this instead: ---
# df <- read.csv("iris.csv")
# --------------------------------------------------------

cat("Dataset shape:", nrow(df), "rows,", ncol(df), "columns\n")
cat("\nFirst 5 rows:\n")
print(head(df, 5))
cat("\nClass distribution:\n")
print(table(df$species))

# ===== Step 2: Prepare Features and Target =====
X <- df[, 1:4]  # sepal_length, sepal_width, petal_length, petal_width
y <- as.factor(df$species)

# ===== Step 3: Split into Training (70%) and Testing (30%) =====
set.seed(42)
train_index <- sample(1:nrow(df), 0.7 * nrow(df))
X_train <- X[train_index, ]
X_test <- X[-train_index, ]
y_train <- y[train_index]
y_test <- y[-train_index]
cat(sprintf("\nTraining samples: %d\n", length(y_train)))
cat(sprintf("Testing samples: %d\n", length(y_test)))

# ===== Step 4: Feature Scaling =====
X_train_scaled <- scale(X_train)
X_test_scaled <- scale(X_test,
                       center = attr(X_train_scaled, "scaled:center"),
                       scale = attr(X_train_scaled, "scaled:scale"))

# ===== Step 5: Build KNN Classifier =====
k <- 5
y_pred <- knn(train = X_train_scaled, test = X_test_scaled, cl = y_train, k = k)
cat(sprintf("\nKNN Model trained and predicted! (K=%d)\n", k))

# ===== Step 6: Evaluate Model Accuracy =====
cm <- confusionMatrix(y_pred, y_test)
cat(sprintf("\n%s\n", strrep("=", 50)))
cat(sprintf("  MODEL ACCURACY: %.2f%%\n", cm$overall['Accuracy'] * 100))
cat(sprintf("%s\n", strrep("=", 50)))
cat("\nFull Evaluation:\n")
print(cm)

# ===== Step 7: Confusion Matrix =====
cm_table <- table(Predicted = y_pred, Actual = y_test)
cat("\nConfusion Matrix:\n")
print(cm_table)

# Plot Confusion Matrix
fourfoldplot(cm_table, color = c("red", "green"),
             main = sprintf("KNN (K=%d) - Confusion Matrix (Iris)", k))

# ===== Step 8: Find Optimal K =====
cat("\nFinding optimal K value...\n")
k_range <- 1:20
accuracies <- numeric(length(k_range))

for (i in seq_along(k_range)) {
  pred <- knn(train = X_train_scaled, test = X_test_scaled, cl = y_train, k = k_range[i])
  accuracies[i] <- sum(pred == y_test) / length(y_test)
}

# Display accuracy for each K
cat(sprintf("\n%-5s %10s\n", "K", "Accuracy"))
cat(sprintf("%-5s %10s\n", strrep("-", 5), strrep("-", 10)))
for (i in seq_along(k_range)) {
  cat(sprintf("%-5d %9.2f%%\n", k_range[i], accuracies[i] * 100))
}

best_k <- k_range[which.max(accuracies)]
cat(sprintf("\nBest K = %d with Accuracy = %.2f%%\n", best_k, max(accuracies) * 100))

# Plot K vs Accuracy
plot(k_range, accuracies, type = "b", col = "blue", pch = 16,
     xlab = "K (Number of Neighbors)", ylab = "Accuracy",
     main = "KNN - Accuracy vs K Value")
abline(v = best_k, col = "red", lty = 2)
legend("bottomright", legend = paste("Best K =", best_k), col = "red", lty = 2)

# ===== Step 9: Predict for New Data =====
new_flower <- data.frame(sepal_length = 5.1, sepal_width = 3.5,
                         petal_length = 1.4, petal_width = 0.2)
new_scaled <- scale(new_flower,
                    center = attr(X_train_scaled, "scaled:center"),
                    scale = attr(X_train_scaled, "scaled:scale"))
prediction <- knn(train = X_train_scaled, test = new_scaled, cl = y_train, k = best_k)
cat(sprintf("\nPrediction for new flower (5.1, 3.5, 1.4, 0.2): %s\n", prediction))